File: bookwyrm/item.py
# coding: utf-8
# This file is part of bookwyrm.
# Copyright 2016, Tmplt.
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

from fuzzywuzzy import fuzz
from collections import namedtuple
from enum import Enum
import argparse
import itertools

FUZZ_RATIO_DEF = 75

NONEXACTS = [
    'authors',
    'title',
    'serie',
    'publisher',
    'journal'
]

EXACTS = [
    'year',
    'lang',
    'edition',
    'ext',
    'volume',
    'number',
    'pages'
]

MISC = [
    'isbns',
    'mirrors'
]


class NonExacts(namedtuple('NonExacts', NONEXACTS)):
    __slots__ = ()

    def __new__(cls, authors, title, serie=None, publisher=None, journal=None):
        return super(NonExacts, cls).__new__(cls, authors, title, serie, publisher, journal)


class Exacts(namedtuple('Exacts', EXACTS)):
    __slots__ = ()

    def __new__(cls, year=None, lang=None, edition=None, ext=None, volume=None, number=None, pages=None):
        return super(Exacts, cls).__new__(cls, year, lang, edition, ext, volume, number, pages)


class Misc(namedtuple('Misc', MISC)):
    __slots__ = ()

    def __new__(cls, isbns=None, mirrors=None):
        return super(Misc, cls).__new__(cls, isbns, mirrors)


class ItemType(Enum):
    paper = 0
    book = 1


class Item:
    """Holds all data for a item."""

    nonexacts = NonExacts
    exacts = Exacts
    misc = Misc

    def __init__(self, namespace=None, nonexacts=None, exacts=None, misc=None):
        if isinstance(namespace, argparse.Namespace):
            self.__init_from_argparse(namespace)
        else:
            self.nonexacts = nonexacts
            self.exacts = exacts
            self.misc = misc

    def __init_from_argparse(self, args):
        self.nonexacts = NonExacts(
            authors = args.author,
            title = args.title,
            serie = args.serie,
            publisher = args.publisher,
            # journal = args.journal
        )

        self.exacts = Exacts(
            # year = args.year,
            # edition = args.edition,
            # ext = args.extension,
            # volume = args.volume,
            # number = args.number
        )

        self.misc = Misc(
            # isbns = args.isbns,
            mirrors = None
        )

    # def __getattr__(self, key):
    #     return getattr(super(Item, self).__getattribute__('data'), key)

    def item_type(self):
        paper_exclusive = (
            self.nonexacts.journal,
            self.exacts.volume,
            self.exacts.number
        )

        if any(paper_exclusive):
            return ItemType.paper

        return ItemType.book

    def matches(self, wanted, fuzzy_min=FUZZ_RATIO_DEF):
        """
        Returns true if all specified exact values are equal
        and if all specified non-exact values pass the fuzzy ratio.
        """

        def match_partial(s1, s2):
            ratio = fuzz.partial_ratio(s1, s2)
            return ratio >= fuzzy_min

        # Don't check if a Paper matches a Book and vise-versa.
        if not self.item_type() is wanted.item_type():
            return False

        # Parallell iteration over the two tuples of exact values.
        for val, req in zip(self.exacts, wanted.exacts):
            if req is not None:
                if val != req:
                    return False

        try:
            if wanted.misc.isbns:
                if not set(wanted.misc.isbns) & set(self.misc.isbns):
                    return False
        except TypeError:
            # Book.isbns exists, but Paper.isbns doesn't.
            pass

        in_result = (self.nonexacts.title, self.nonexacts.serie, self.nonexacts.publisher)
        requested = (wanted.nonexacts.title, wanted.nonexacts.serie, wanted.nonexacts.publisher)

        for val, req in zip(in_result, requested):
            if req is not None:
                # partial: useful for course literature which can have some
                # crazy long titles. Also useful for publisher,
                # because some entries may not use the full name of it.
                if not match_partial(val, req):
                    return False

        # token_set seems to work best here, both when only
        # the last name is given but also when something like
        # "J. Doe" is given.
        if wanted.nonexacts.authors:
            ratio_thus_far = 0
            for comb in itertools.product(self.nonexacts.authors, wanted.nonexacts.authors):
                ratio = fuzz.token_set_ratio(comb[0], comb[1])
                ratio_thus_far = max(ratio, ratio_thus_far)

            if ratio_thus_far < fuzzy_min:
                return False

        return True

File: src/core/item.cpp
#include <cctype>

#include <fuzzywuzzy.hpp>

#include "../string.hpp"
#include "algorithm.hpp"
#include "common.hpp"
#include "item.hpp"

namespace bookwyrm::core {

    size_t item::items_idx = 0;

    static int get_integral(const py::dict &dict, const char *key)
    {
        if (!dict.contains(key))
            return empty;
        return py::int_(dict[key]);
    }

    static string get_string(const py::dict &dict, const char *key)
    {
        if (!dict.contains(key))
            return "";
        return trim(py::str(dict[key]));
    }

    static vector<string> get_vector_string(const py::dict &dict, const char *key)
    {
        if (!dict.contains(key))
            return {{}};

        vector<string> strings;
        for (auto &handle : py::list(dict[key]))
            strings.push_back(trim(py::str(handle)));
        return strings;
    }

    exacts_t::exacts_t(const py::dict &dict)
        : ymod(year_mod::unused), year(get_integral(dict, "year")), volume(get_integral(dict, "volume")),
          number(get_integral(dict, "number")), pages(get_integral(dict, "pages")), size(get_integral(dict, "size")),
          extension(get_string(dict, "extension"))
    {
    }

    nonexacts_t::nonexacts_t(const py::dict &dict)
        : authors(get_vector_string(dict, "authors")), title(get_string(dict, "title")), series(get_string(dict, "series")),
          publisher(get_string(dict, "publisher")), journal(get_string(dict, "journal")),
          edition(get_string(dict, "edition"))
    {
    }

    misc_t::misc_t(const py::dict &dict) : uris(get_vector_string(dict, "uris")), isbns(get_vector_string(dict, "isbns")) {}

    bool exacts_t::operator==(const exacts_t &other) const
    {
        /* extension is checked first, because some sources offer items in multiple
         * formats. */
        return std::tie(extension, ymod, year, volume, number, pages, size) ==
               std::tie(other.extension, other.ymod, other.year, other.volume, other.number, other.pages, other.size);
    }

    bool nonexacts_t::operator==(const nonexacts_t &other) const
    {
        return std::tie(authors, title, series, publisher, journal, edition) ==
               std::tie(other.authors, other.title, other.series, other.publisher, other.journal, other.edition);
    }

    bool misc_t::operator==(const misc_t &other) const { return std::tie(uris, isbns) == std::tie(other.uris, other.isbns); }

    bool item::operator==(const item &other) const
    {
        return std::tie(nonexacts, exacts, misc) == std::tie(other.nonexacts, other.exacts, other.misc);
    }

    bool item::matches(const item &wanted, const unsigned int fuzzy_min) const
    {
        /* Return false if any exact value doesn't match what's wanted. */
        for (const auto & [ req, got ] : func::zip(wanted.exacts.store, this->exacts.store)) {
            if (req != empty && req != got)
                return false;
        }

        if (wanted.exacts.year != empty) {
            const auto &req = wanted.exacts.year, got = this->exacts.year;

            switch (wanted.exacts.ymod) {
            case year_mod::equal:
                if (!(got == req))
                    return false;
                break;
            case year_mod::eq_gt:
                if (!(got >= req))
                    return false;
                break;
            case year_mod::eq_lt:
                if (!(got <= req))
                    return false;
                break;
            case year_mod::gt:
                if (!(got > req))
                    return false;
                break;
            case year_mod::lt:
                if (!(got < req))
                    return false;
                break;
            case year_mod::unused:
                /* Should never happen. */
                assert(false);
            }
        }

        /* Ad-hoc the file type, for now. */
        if (!wanted.exacts.extension.empty()) {
            if (this->exacts.extension != wanted.exacts.extension)
                return false;
        }

        /* Does the item contain a wanted ISBN? */
        if (!wanted.misc.isbns.empty() && !func::any_intersection(wanted.misc.isbns, this->misc.isbns))
            return false;

        const std::array<string, 3> in_result = {{this->nonexacts.title, this->nonexacts.series, this->nonexacts.publisher}},
                                    requested = {
                                        {wanted.nonexacts.title, wanted.nonexacts.series, wanted.nonexacts.publisher}};

        for (const auto & [ req, got ] : func::zip(requested, in_result)) {
            if (!req.empty()) {
                /*
                 * partial: useful for course literature that can have some
                 * crazy long titles. Also useful for publishers, because
                 * some entries may not use the full name.
                 */
                if (fuzz::partial_ratio(got, req) < fuzzy_min)
                    return false;
            }
        }

        if (!wanted.nonexacts.authors.empty()) {
            unsigned int max_ratio = 0;
            for (const auto & [ req, got ] : algorithm::product(wanted.nonexacts.authors, this->nonexacts.authors)) {
                /*
                 * From some quick testing, it feels like token_set_ratio
                 * works best here.
                 */
                auto ratio = fuzz::token_set_ratio(req, got);
                max_ratio = std::max(ratio, max_ratio);

                if (max_ratio >= fuzzy_min)
                    break;
            }

            if (max_ratio < fuzzy_min)
                return false;
        }

        return true;
    }

} // namespace bookwyrm::core
